### PR TITLE
Model Refactor 2 - Model Project + Webapp Model

### PR DESCRIPTION
**Pull request by IstvanM**
I've introduced the Reference object in the model.py. 
This is needed as we store objects as "Documents" (Pickled Objects). This will be ideal for DynamoDB, MongoDB, Redis etc. as Storage. 

However, as python pickling would store a copy of a Project and Connection to every Dashboard or SavedMetric. 
Changes made to the original Project or Connection wouldn't change the ones in Dashboards or SavedMetrics.

The Reference class is a compound of an Object and its ID (the is always equal to the Objects own ID). 
The ID itself is pickled and the object is not. 

During object retrieval from Storage only the ID is restored to the Reference so we manually need to get the Object (for example the Project). This will ensure that there are no copies of Project and Connection in every dashboard.

### CODE REVIEW

File: mitzu/model.py [hunk 1]
```diff
@@ -322,7 +323,17 @@ def to_relative_delta(self) -> relativedelta:
         raise Exception(f"Unsupported relative delta value: {self.period}")
 
 
+class Identifiable(Protocol):
```
Comment (IstvanM) [inline mitzu/model.py:326]: This only ensures that some high level classes will have an ID. These classes are stored in the Storage.

File: mitzu/model.py [hunk 2]
```diff
@@ -678,13 +717,23 @@ def validate(self, adapter: GA.GenericDatasetAdapter):
                     f"Event data table '{self.table_name}' does not have '{field_to_validate._get_name()}' field"
                 )
 
+    @property
```
Comment (IstvanM) [inline mitzu/model.py:720]: Property in python is just a method that doesn't need to be called with parenthesis...

File: mitzu/model.py [hunk 3]
```diff
@@ -803,6 +871,7 @@ def _to_globals(cls, glbs: Dict):
 class DiscoveredProject:
```
Comment (IstvanM) [inline mitzu/model.py:871]: Somehow I feel that the DiscoveredProject should disappear.

File: mitzu/webapp/model.py [hunk 4]
```diff
@@ -7,29 +7,82 @@
...
+    ):
 
-@dataclass()
+        object.__setattr__(self, "chart", chart)
+        object.__setattr__(self, "image_base64", image_base64)
+        object.__setattr__(self, "small_base64", small_base64)
+        object.__setattr__(self, "saved_at", datetime.now())
+        object.__setattr__(self, "_project_ref", M.Reference(project))
+        object.__setattr__(self, "_metric_state", M.State(metric))
+        object.__setattr__(self, "_metric_json", SE.to_compressed_string(metric))
```
Comment (IstvanM) [inline mitzu/webapp/model.py:53]: The metric class can't be pickled or unpickled as the Project is stored very deep in the EventDataTables.  So for now I am storing our own Serialized json version of the Metric.

File: mitzu/webapp/model.py [hunk 5]
```diff
@@ -7,29 +7,82 @@
...
+            raise M.InvalidReferenceException(
+                "Project is missing from SavedMetric reference"
+            )
+        return res
+
+    @property
+    def metric(self) -> M.Metric:
+        res = self._metric_state.get_value()
+        if res is None:
+            res = SE.from_compressed_string(self._metric_json, self.project)
```
Comment (IstvanM) [inline mitzu/webapp/model.py:68]: Here we resolve the Metric of the SavedMetric itself in a lazy way.

File: mitzu/webapp/model.py [hunk 6]
```diff
@@ -42,23 +95,58 @@ class DashboardMetric:
...
-    x: int = 0
-    y: int = 0
-    width: int = 2
-    height: int = 8
-    saved_metric: Optional[SavedMetric] = None
+    x: int
+    y: int
+    width: int
+    height: int
+    _saved_metric_ref: M.Reference[SavedMetric]
```
Comment (IstvanM) [inline mitzu/webapp/model.py:102]: Same as for the Project and Connection. The DashboardMetric should only hold a reference to the SavedMetrics.

**Review by medgyes — APPROVED** [review]

**Pull request merged**